_posts/success_stories/2024-05-20-lavinmq-understanding-code.md

```diff
@@ -0,0 +1,50 @@
+
+[LavinMQ](https://lavinmq.com/) was built by experienced developers from 84codes, founders of one of the largest [cloud messaging platforms](https://www.cloudamqp.com/), CloudAMQP. By prioritizing the insurance of stable communication and reliable message delivery in and between services, LavinMQ was built to process data quickly and efficiently while allowing decoupled system components to operate independently and scale as needed.
+
+LavinMQ can handle numerous connections between the publisher, the queue, and the consumer, which is often a requirement in e.g., IoT architectures. It’s an open-source message broker, pre-optimized to be extremely fast compared to other brokers, and can easily handle 1 million messages per second.
+
+LavinMQ includes all the features people want the most and leaves just a small memory footprint, supporting:
+
+* Streaming
+* Flexible Routing
+* AMQP 0.9.1 Protocol
+* Acknowledgments and Confirms
+* Stream Queue
+* Unlimited Queue Length (dependent on disk space)
+* Replication
+
+We highlighted the characteristics that stand out from a technical point of view. How does it manage to handle more than a million messages per second, and have a low memory footprint? Half of it is due to, and we quote:
+
+> Messages are written to the disk really fast and the caching is managed by the OS itself. This leads to a performance boost while keeping RAM footprint low.
+
+But the other half, as mentioned in [their blog](https://lavinmq.com/blog/crystal-clear-message-brokering), is due to it being developed in Crystal. And this comes in several interesting ways.
+
+1. Crystal is efficiently compiled thanks to its LLVM backend. This enables aggressive optimizations, like inlining methods and removing indirections.
+
+2. Crystal helps developers allocate their data either on the heap or on the stack. Heap allocations are expensive, so allocating data on the stack in the hot-path of the program brings an important boost in performance. Crystal makes it easy to choose where data should be stored.
+
+3. Crystal's stdlib and compiler are written in Crystal itself. One of Crystal's golden characteristics is its readability, making it possible to read the code that is going to be run by the program without having to resort to another, typically lower-level  language.
+
+The latter is the main point of this post. In fact, the documentation of the standard library points to the code, making the performance implications of calling a certain method explicit. As an example that will come up later, the documentation for the `file` method of the `Digest` class can be found [here](https://crystal-lang.org/api/1.12.1/Digest.html#file%28file_name%3APath%7CString%29%3Aself-instance-method). Clicking on [[View source](https://github.com/crystal-lang/crystal/blob/4cea10199/src/digest/digest.cr#L214)] we land in the method’s code.

```

Comment:
**issue:** *here* links are bad. Link texts should be descriptive, and this is easy to achieve here.
I also took the liberty to shorten the label to the formal method reference instead of a verbal description for conciseness.

"As an example that will come up later," doesn't make sense here. We're directly using that example right now.

```suggestion
The latter is the main point of this post. In fact, the documentation of the standard library points to the code, making the performance implications of calling a certain method explicit. For example, look at the [documentation for `Digest.file`](https://crystal-lang.org/api/1.12.1/Digest.html#file%28file_name%3APath%7CString%29%3Aself-instance-method). Clicking on [*View source*](https://github.com/crystal-lang/crystal/blob/4cea10199/src/digest/digest.cr#L214) we land in the method’s code.

```